app/services/remove_cps_service.rb

```diff
@@ -0,0 +1,58 @@
+class RemoveCpsService
+  def call
+    subscriber_lists = SubscriberList.where('content_purpose_supergroup IS NOT NULL')
+    subscriber_lists.each do |subscriber_list|
+      matching_list = matching_subscriber_list(subscriber_list)
+      if matching_list.nil?
+        content_purpose_supergroup_to_tags(subscriber_list)
+      else
+        move_matched_content_changes(subscriber_list, matching_list)
+        move_subscriptions(subscriber_list, matching_list)
+        subscriber_list.destroy
+      end
+    end
+  end
+
+private
+
+  def active_subscribers(subscriber_list)
+    Subscriber.joins(:subscriptions).where(subscriptions: { subscriber_list: subscriber_list, ended_at: nil })
+  end
+
+  def move_subscriptions(from_subscriber_list, to_subscriber_list)
+    subscriptions = from_subscriber_list.subscriptions
+    subscribers = active_subscribers(to_subscriber_list)
+    subscriptions.each do |subscription|
+      if subscription.active? && subscribers.include?(subscription.subscriber)
+        subscription.end(reason: 'subscriber_list_changed')
+      end
+      subscription.update_attribute(:subscriber_list, to_subscriber_list)
+    end
+  end
+
+  def move_matched_content_changes(from_subscriber_list, to_subscriber_list)
```

Comment:
What happens if a new content change comes in while this function is running?

Reply:
That is also a good point.

I changed the code so that the process is now in two steps. One deactivates the legacy subscriber lists by moving all the subscriptions. Afterwards, there should not be new matched_content_changes that affect these lists.

The second rake task removes the superfluous lists and move the associated matched_content_changes.